flink-connectors/flink-connector-gcp-pubsub/src/main/java/org/apache/flink/streaming/connectors/gcp/pubsub/PubSubSink.java

```diff
@@ -0,0 +1,272 @@
+	/**
+	 * Create a builder for a new PubSubSink.
+	 *
+	 * @param <IN> The generic of the type that is to be written into the sink.
+	 * @return a new PubSubSinkBuilder instance
+	 */
+	public static <IN> PubSubSinkBuilder<IN> newBuilder(SerializationSchema<IN> serializationSchema, String projectName, String topicName) {
+		return new PubSubSinkBuilder<>(serializationSchema, projectName, topicName);
+	}
+
+	@Override
+	public void snapshotState(FunctionSnapshotContext context) throws Exception {
+		//before checkpoints make sure all the batched / buffered pubsub messages have actually been sent
+		publisher.publishAllOutstanding();
+	}
+
+	@Override
+	public void initializeState(FunctionInitializationContext context) throws Exception {
+	}
+
+	/**
+	 * PubSubSinkBuilder to create a PubSubSink.
+	 *
+	 * @param <IN> Type of PubSubSink to create.
+	 */
+	@SuppressWarnings("unchecked")
+	public static class PubSubSinkBuilder<IN> {
+		private SerializationSchema<IN> serializationSchema;
+		private String projectName;
+		private String topicName;
+
+		private Credentials credentials;
+		private String hostAndPort;
+
+		private PubSubSinkBuilder(SerializationSchema<IN> serializationSchema, String projectName, String topicName) {
+			this.serializationSchema = serializationSchema;
+			this.projectName = projectName;
+			this.topicName = topicName;
+		}
+
+		/**
+		 * Set the credentials.
+		 * If this is not used then the credentials are picked up from the environment variables.
+		 *
+		 * @param credentials the Credentials needed to connect.
+		 * @return The current PubSubSinkBuilder instance
+		 */
+		public PubSubSinkBuilder<IN> withCredentials(Credentials credentials) {
+			this.credentials = credentials;
+			return this;
+		}
+
+		/**
+		 * @param serializationSchema Instance of a SerializationSchema that converts the IN into a byte[]
+		 * @return The current PubSubSinkBuilder instance
+		 */
+		public PubSubSinkBuilder<IN> withSerializationSchema(SerializationSchema<IN> serializationSchema) {
+			this.serializationSchema = serializationSchema;
+			return this;
+		}
+
+		/**
+		 * @param projectName The name of the project in PubSub
+		 * @return The current PubSubSinkBuilder instance
+		 */
+		public PubSubSinkBuilder<IN> withProjectName(String projectName) {
+			this.projectName = projectName;
+			return this;
+		}
+
+		/**
+		 * @param topicName The name of the topic in PubSub
+		 * @return The current PubSubSinkBuilder instance
+		 */
+		public PubSubSinkBuilder<IN> withTopicName(String topicName) {
+			this.topicName = topicName;
+			return this;
+		}
+
+		/**
+		 * Set the custom hostname/port combination of PubSub.
+		 * The ONLY reason to use this is during tests with the emulator provided by Google.
+		 *
+		 * @param hostAndPort The combination of hostname and port to connect to ("hostname:1234")
+		 * @return The current PubSubSinkBuilder instance
+		 */
+		public PubSubSinkBuilder<IN> withHostAndPortForEmulator(String hostAndPort) {
```

Comment:
Personally I find a method like this very useful when running integration tests. If we would make this package prive it would not be possible for end users to use the connector in integration tests.

I'd like to keep it as-is if you agree